1.问题:编写一个可以计算给定数的阶乘的程序。结果应该以逗号分隔的顺序打印在一行上。例如：输入8，则输出8的阶乘40320

In [2]:
def factorial(n):
    result = 1
    for i in range(n):
        result *= (i+1)
    return result
factorial(int(input()))

40320

2.问题:编写一个程序，它将找到1000至2200(包括在内)所有的可被5整除，但不是7的倍数的数字。得到的数字应按逗号分隔的顺序打印在一行上。

In [6]:
def find5(**kwargs):
    start, end = 1000, 2200
    for i in range(start, end+1, 5):
        if i%7:
            yield i
for i in find5():
    if i==2200:
        print(i, end='')
    else:
        print(i, end=',')

1000,1005,1010,1020,1025,1030,1035,1040,1045,1055,1060,1065,1070,1075,1080,1090,1095,1100,1105,1110,1115,1125,1130,1135,1140,1145,1150,1160,1165,1170,1175,1180,1185,1195,1200,1205,1210,1215,1220,1230,1235,1240,1245,1250,1255,1265,1270,1275,1280,1285,1290,1300,1305,1310,1315,1320,1325,1335,1340,1345,1350,1355,1360,1370,1375,1380,1385,1390,1395,1405,1410,1415,1420,1425,1430,1440,1445,1450,1455,1460,1465,1475,1480,1485,1490,1495,1500,1510,1515,1520,1525,1530,1535,1545,1550,1555,1560,1565,1570,1580,1585,1590,1595,1600,1605,1615,1620,1625,1630,1635,1640,1650,1655,1660,1665,1670,1675,1685,1690,1695,1700,1705,1710,1720,1725,1730,1735,1740,1745,1755,1760,1765,1770,1775,1780,1790,1795,1800,1805,1810,1815,1825,1830,1835,1840,1845,1850,1860,1865,1870,1875,1880,1885,1895,1900,1905,1910,1915,1920,1930,1935,1940,1945,1950,1955,1965,1970,1975,1980,1985,1990,2000,2005,2010,2015,2020,2025,2035,2040,2045,2050,2055,2060,2070,2075,2080,2085,2090,2095,2105,2110,2115,2120,2125,2130,2140,2145,2150,2155,2160,

3.问题：有四个数字：1、2、3、4、5，能组成多少个互不相同的三位数？各是多少？

In [9]:
def compose():
    count = 0
    for a in range(1, 6):
        for b in range(1, 6):
            if a!=b:
                for c in range(1, 6):
                    if a!=c and b!=c:
                        count += 1
                        print('{}{}{}'.format(a, b, c), end='\t')
                    else:
                        continue
            else:
                continue
    else:
        print('')
        print('总共有%d个三位数。' % count)
compose()

123	124	125	132	134	135	142	143	145	152	153	154	213	214	215	231	234	235	241	243	245	251	253	254	312	314	315	321	324	325	341	342	345	351	352	354	412	413	415	421	423	425	431	432	435	451	452	453	512	513	514	521	523	524	531	532	534	541	542	543	
总共有60个三位数。


4.问题：一个整数，它加上100后是一个完全平方数，再加上168又是一个完全平方数，请问该数是多少？

In [17]:
import math
def guess_number():
    ls = [i for i in range(1, 21)]
    ls = list(map(lambda x:math.pow(x, 2), ls))
    left, right = 0, 0
    for i in range(20):
        for j in range(i, 20):
            if ls[j]-ls[i]==68:
                print('这个数是：{}'.format(int(ls[i]-100)))
                break
guess_number()

这个数是：156


5.问题：输入某年某月某日，判断这一天是这一年的第几天？例如：1998年3月5日，应该先把前两个月的天数加起来，然后再加上5天即本年的第几天

In [22]:
def is_leap_year(year):
    return (year % 4 == 0 and year % 100 != 0) or year % 400 == 0

def count_day(date):
    ls = date.split('年')
    temp = ls.pop(1)[:-1].split('月')
    y, m, d = map(int, ls + temp)
    day_list = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    if is_leap_year(y):
        day_list[1] = 29
    return sum(day_list[:m]) + d

count_day(input())

1998年3月5日


95

6.问题：将一个正整数分解质因数。例如：输入90,打印出90=2\*3\*3\*5。

In [52]:
import math
def prime(num, prime_list=[]):
    for i in range(2, math.ceil(math.sqrt(num))+1):
        if num%i==0:
            prime_list.append(i)
            prime(num/i, prime_list)
            return prime_list
        if i>=math.ceil(math.sqrt(num))-1:
            prime_list.append(int(num))
            break
n = int(input('待分解质因数的整数：'))
result = prime(n)
print(n, end='=')
length = len(result)
for i in range(length):
    if length-1==i:
        print(result[i])
    else:
        print(result[i], end='*')

待分解质因数的整数：90
90=2*3*3*5


7.问题：输入一行字符，分别统计出其中英文字母、空格、数字和其它字符的个数。例如：输入Hello world! 123 输出应该是：字母10 数字3 符号1.

In [3]:
string = input()
count_alpha, count_num, count_char, count_space = 0, 0, 0, 0
for i in string:
    if i.isalpha():
        count_alpha+=1
    elif i.isdigit():
        count_num+=1
    elif i.isspace():
        count_space+=1
    else:
        count_char+=1
if count_alpha:
    print("字母{}".format(count_alpha), end=' ')
if count_space:
    print("空格{}".format(count_space), end=' ')
if count_num:
    print("数字{}".format(count_num), end=' ')
if count_char:
    print("符号{}".format(count_char), end=' ')

Hello world! 123
字母10 空格2 数字3 符号1 

8.问题：给一个不多于5位的正整数，要求：一、求它是几位数，二、逆序打印出各位数字.

In [22]:
def count_and_reverse(num):
    return len(list(reversed([i for i in str(num)]))), ''.join(list(reversed([i for i in str(num)])))
count_and_reverse(12345)

(5, '54321')

9.问题：利用递归函数调用方式，将所输入的5个字符，以相反顺序打印出来。

In [23]:
def reverse_by_recursion(string):
    if len(string)==1:
        print(string, end='')
    else:
        reverse_by_recursion(string[1:])
        print(string[0], end='')
reverse_by_recursion(input())

hello
olleh

10.问题：编写一个函数，输入n为偶数时，调用函数求1/2+1/4+...+1/n,当输入n为奇数时，调用函数1/1+1/3+...+1/n

In [1]:
def sum_(n):
    def single_(n):
        cursor = 1
        while cursor<=n:
            yield 1/cursor
            cursor+=2
    def double_(n):
        cursor = 2
        while cursor<=n:
            yield 1/cursor
            cursor+=2
    if n%2:
        return sum(single_(n))
    else:
        return sum(double_(n))
print(sum_(6))

0.9166666666666666


11.问题：计算字符串中子串出现的次数。例如输入：字符串：abcbab，字串：ab，输出2.

In [3]:
def count_sub(string, sub_string):
    return string.count(sub_string)
count_sub('abcabc', 'ab')

2

12.问题：输入一个奇数，然后判断最少几个 9 除于该数的结果为整数。例如：输入13 ，则输出999999； 999999 / 13 = 76923

In [8]:
def how_many(num):
    if num%2:
        nine = 9
        while True:
            if nine%num:
                nine = nine*10 + 9
            else:
                return nine
    else:
        print("输入值要求必须为奇数")
num = int(input())
result = how_many(num)
print("{} / {} = {:.0f}".format(result, num, result/num))

13
999999 / 13 = 76923


13.问题：从键盘输入一些字符，逐个把它们写到磁盘文件上，直到输入一个 # 为止：

In [5]:
import winreg
def get_desktop():
    key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, 'Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\Shell Folders')
    return winreg.QueryValueEx(key, 'Desktop')[0]
desktop_path = get_desktop()
print("请输入字符，文件将保存于桌面的 file.txt 内, 输入 # 结束输入")
with open(desktop_path+'\\file.txt', 'a', encoding='utf-8') as f:
    s = input()
    while s!='#':
        f.write(s+'\n')
        s = input()

请输入字符，文件将保存于桌面的 file.txt 内, 输入 # 结束输入
helloworld
helloagain
#


14.问题：编写一个程序，接受一行序列作为输入，并在将句子中的所有字符大写后打印出来，例如：输入Hello world，输出HELLO WORLD.

In [7]:
s = input()
print(s.upper())

helloworld
HELLOWORLD


15.问题：对10个数进行排序。选择排序和冒泡排序两种方法。

In [19]:
def select_sort(ls):
    left, right = 0, 9
    while left<right:
        min_index = left
        max_index = right
        for i in range(left, right+1):
            if ls[i]<ls[min_index]:
                min_index = i
            if ls[i]>ls[max_index]:
                max_index = i
        
        if ls[min_index]<ls[left]:
            ls[min_index], ls[left] = ls[left], ls[min_index]
        if max_index==left:
            max_index = min_index
        if ls[max_index]>ls[right]:
            ls[max_index], ls[right] = ls[right], ls[max_index]
        
        left+=1
        right-=1
    else:
        return ls
    print("发生了预料之外的错误")
import random
ls = []
for i in range(10):
    ls.append(random.randint(0, 100))
print('排序前：', ls)
print('排序后：', select_sort(ls))
if sorted(ls)==select_sort(ls):
    print('排序结果正确')
else:
    print('排序出错')

排序前： [25, 86, 78, 73, 17, 83, 21, 36, 31, 13]
排序后： [13, 17, 21, 25, 31, 36, 73, 78, 83, 86]
排序结果正确


In [21]:
def bubble_sort(ls):
    left, right = 0, 9
    while left<right:
        if ls[left]>ls[left+1]:
            ls[left], ls[left+1] = ls[left+1], ls[left]
        if ls[right]<ls[right-1]:
            ls[right], ls[right-1] = ls[right-1], ls[right]
        left+=1
        right-=1
    return ls
import random
ls = []
for i in range(10):
    ls.append(random.randint(0, 100))
print('排序前：', ls)
print('排序后：', select_sort(ls))
if sorted(ls)==select_sort(ls):
    print('排序结果正确')
else:
    print('排序出错')

排序前： [100, 38, 58, 63, 79, 81, 6, 48, 65, 68]
排序后： [6, 38, 48, 58, 63, 65, 68, 79, 81, 100]
排序结果正确


In [3]:
import functools
reduce(8)

NameError: name 'reduce' is not defined